crates/xtask/src/changelog.rs

```diff
@@ -0,0 +1,549 @@
+                                .expect("No last description to append to.");
+                            last.push_str(format!("\n{current_line}").as_str());
+                            release_descriptions.push(last);
+                        }
+                    }
+
+                    match release_type {
+                        ReleaseType::Major => release.major = release_descriptions,
+                        ReleaseType::Minor => release.minor = release_descriptions,
+                        ReleaseType::Patch => release.patch = release_descriptions,
+                    }
+                }
+
+                // Validate descending versions.
+                if let Some(previous_version) = releases.last() {
+                    if release.version > previous_version.version {
+                        return Err(anyhow!("Versions out of order"));
+                    }
+                }
+
+                // Validate pre-release.
+                if !releases.is_empty() && !release.version.pre.is_empty() {
+                    return Err(anyhow!("Only the first version can be pre-release"));
+                }
+
+                releases.push(release);
+            }
+
+            // Hit last line. Done parsing releases. Look for skip counter.
+            if next_line.is_none() {
+                skip_counter = current_line
+                    .trim_start_matches("<!-- Increment to skip CHANGELOG.md test: ")
+                    .trim_end_matches(" -->")
+                    .parse::<u32>()
+                    .map_err(|err| {
+                        anyhow!("Invalid skip_counter at the end of the changelog ({err})")
+                    })?;
+            }
+        }
+
+        if let Some(last_release) = releases.last() {
+            if last_release.version.to_string() != "0.1.0" {
+                return Err(anyhow!("The last release must be version 0.1.0"));
+            }
+
+            if last_release.major.len() > 0
+                || last_release.minor.len() > 0
+                || last_release.patch.len() > 0
+            {
+                return Err(anyhow!("The last release must contain no changes"));
+            }
+        } else {
+            return Err(anyhow!("At least 1 release is required"));
+        }
+
+        Ok(Changelog { releases, skip_counter })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct Release {
+    version: Version,
+
+    major: Vec<String>,
+    minor: Vec<String>,
+    patch: Vec<String>,
+}
+
+impl Release {
+    fn from(version: &str) -> Result<Self> {
+        let semver = Version::parse(version)?;
+
+        Ok(Release { version: semver, major: Vec::new(), minor: Vec::new(), patch: Vec::new() })
+    }
+}
+
+/// Validates CHANGELOG.md files for all Wasefire crates.
+pub fn execute_ci() -> Result<()> {
+    let dir_entries = wasefire_cli_tools::fs::read_dir("crates")?;
+
+    let existing_changelog_paths = dir_entries.filter_map(|entry| {
```

Comment:
Since you have a for loop just below, I would inline the filtering logic there using `continue`.